### PR TITLE
Support atomic bit writing protocol, instead of read-modify-write

### PR DESCRIPTION
**Pull request by sam0737**
PR for #76

### CODE REVIEW

File: S7.Net/PLC.cs [hunk 1]
```diff
@@ -540,14 +540,19 @@ public void ReadClass(object sourceClass, int db, int startByteAdr = 0)
         /// <param name="startByteAdr">Start byte address. If you want to write DB1.DBW200, this is 200.</param>
         /// <param name="value">Bytes to write. If more than 200, multiple requests will be made.</param>
         /// <returns>NoError if it was successful, or the error is specified</returns>
-        public ErrorCode WriteBytes(DataType dataType, int db, int startByteAdr, byte[] value)
+        public ErrorCode WriteBytes(DataType dataType, int db, bool bitMode, int startByteAdr, int bitAdr, byte[] value)
```
Comment (mesta1) [inline S7.Net/PLC.cs]: It is better to write a separate function for WriteBit, and keep WriteBytes as it is. 
WriteBytes writes bytes, WriteBit writes bit.
WriteBit(int db, int startByteAdr, int bitAdr, bool value)

File: S7.Net/PLC.cs [hunk 2]
```diff
@@ -540,14 +540,19 @@ public void ReadClass(object sourceClass, int db, int startByteAdr = 0)
...
         {
             int localIndex = 0;
             int count = value.Length;
+            if (bitMode && value.Length != 1)
```
Comment (mesta1) [inline S7.Net/PLC.cs]: With separate functions, this check ins not more needed, because value becomes a boolean.

File: S7.Net/PLC.cs [hunk 3]
```diff
@@ -684,14 +694,8 @@ public ErrorCode Write(string variable, object value)
...
-                                }
-                                byte b = (byte)Read(DataType.DataBlock, mDB, mByte, VarType.Byte, 1);
-                                if (Convert.ToInt32(value) == 1)
-                                    b = (byte)(b | (byte)Math.Pow(2, mBit)); // Bit setzen
-                                else
-                                    b = (byte)(b & (b ^ (byte)Math.Pow(2, mBit))); // Bit rücksetzen
-
-                                return Write(DataType.DataBlock, mDB, mByte, (byte)b);
+                                }
+                                return Write(DataType.DataBlock, mDB, mByte, (bool)value, mBit);
```
Comment (mesta1) [inline S7.Net/PLC.cs]: Need to check if the method before was working with true/false or with 0/1.

File: S7.Net/Types/Boolean.cs [hunk 4]
```diff
@@ -8,31 +8,12 @@ namespace S7.Net.Types
     public static class Boolean
     {
         /// <summary>
-        /// Returns the value of a bit in a bit, given the address of the bit
+        /// Converts a bit to byte array
         /// </summary>
-        public static bool GetValue(byte value, int bit)
```
Comment (mesta1) [inline S7.Net/Types/Boolean.cs]: These are all public methods. We can't just remove these, but we have to mark the old one as obsolete and add the new ones.

File: S7.Net/PLC.cs (hunk 1 above)
Comment (sam0737) [inline S7.Net/PLC.cs]: There are so many magic numbers for the TPKT/COTP/S7 protocol that I am not exactly comfortable in splitting it.

File: S7.Net/PLC.cs (hunk 2 above)
Comment (sam0737) [inline S7.Net/PLC.cs]: I actually don't know for sure how does writing multiple bits work.

File: S7.Net/Types/Boolean.cs (hunk 4 above)
Comment (sam0737) [inline S7.Net/Types/Boolean.cs]: Ya I can keep them.

File: S7.Net/PLC.cs (hunk 1 above)
Comment (mesta1) [inline S7.Net/PLC.cs]: This is the public WriteBytes, which doesn't contain any code of the protocol. 
The method with magic numbers is WriteBytesWithASingleRequest, which is fine.

**Comment by mesta1** [comment]
Let me know if you have time to update the PR, I would really like to merge these changes into the codebase.

**Comment by sam0737** [comment]
Sorry I am on something else so might not be available to make these changes recently.

**Comment by uknickwilliams** [comment]
Is this pull request still waiting on these changes? If so can I pick them up?

**Comment by mesta1** [comment]
Sure, go on.

**Comment by uknickwilliams** [comment]
How do I push my changes to submit a pull request? I am not allowed to push a new branch to the repository. Get a 403 error.

**Comment by mesta1** [comment]
Fork the repo in your account, create a branch, push your code and create a commit, then you can create a PR.

**Comment by mesta1** [comment]
Merged as part of https://github.com/killnine/s7netplus/pull/89

**Pull request closed without merging**